File: 199/binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> ans;
    vector<int> rightSideView(TreeNode* root) {
        rightSideView(root, 0);
        return ans;
    }
    void rightSideView(TreeNode* root, int level) {
        if(root==nullptr) return;

        if(level>=ans.size()){
            ans.push_back(root->val);
        }else{
            ans[level] = root->val;
        }
        rightSideView(root->left, level+1);
        rightSideView(root->right, level+1);
    }
};

File: 199/binary-tree-right-side-view.py
# LeetCode 199. Binary Tree Right Side View
# 將 binary tree 橫向壓扁，每層的「最右邊」串起來、當答案
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        def helper(root, level):  # 利用「函式呼叫函式」解 tree 問題
            if root==None: return  # 遇到空的末稍，就結束
            if len(ans)<=level:  # 如果 ans 層數不夠
                ans.append(root.val)  # 就新增一層
            else:  # 如果層數夠
                ans[level] = root.val  # 就持續更新
            helper(root.left, level+1)  # 函式呼叫函式
            helper(root.right, level+1)  # 函式呼叫函式
        helper(root, 0)  # 啟動「函式呼叫函式」
        return ans
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
